ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java

```diff
@@ -61,82 +65,62 @@
 
-    this.streamsException = new AtomicReference<>(null);
     this.queryMetadata.getKafkaStreams().setUncaughtExceptionHandler(new StreamsExceptionHandler());
-
     queryMetadata.getKafkaStreams().start();
   }
 
   @Override
   public void write(OutputStream out) throws IOException {
     try {
-      AtomicBoolean rowsWritten = new AtomicBoolean(false);
-      QueryRowWriter queryRowWriter = new QueryRowWriter(
-          out,
-          streamsException,
-          queryMetadata.getRowQueue(),
-          rowsWritten
-      );
-      Thread rowWriterThread = new Thread(queryRowWriter);
-      rowWriterThread.start();
-      try {
-        while (true) {
-          Thread.sleep(disconnectCheckInterval);
-          Throwable exception = streamsException.get();
-          if (exception != null) {
-            throw exception;
-          }
+      while (true) {
+
+        KeyValue<String, GenericRow> value = queryMetadata.getRowQueue().poll(disconnectCheckInterval, TimeUnit.MILLISECONDS);
+        if (value != null) {
+          write(out, value.value);
+        } else {
           // If no new rows have been written, the user may have terminated the connection without
           // us knowing. Check by trying to write a single newline.
-          if (!rowsWritten.getAndSet(false)) {
-            synchronized (out) {
-              out.write("\n".getBytes());
-              out.flush();
-            }
-          }
-        }
-      } catch (EOFException exception) {
-        // The user has terminated the connection; we can stop writing
-      } catch (InterruptedException exception) {
-        // The most likely cause of this is the server shutting down. Should just try to close
-        // gracefully, without writing any more to the connection stream.
-        log.warn("Interrupted while writing to connection stream");
-      } catch (Throwable exception) {
-        log.error("Exception occurred while writing to connection stream: ", exception);
-        synchronized (out) {
-          out.write("\n".getBytes());
-          if (exception.getCause() instanceof KsqlException) {
-            new ObjectMapper().writeValue(out, new StreamedRow(exception.getCause()));
-          } else {
-            new ObjectMapper().writeValue(out, new StreamedRow(exception));
-          }
           out.write("\n".getBytes());
           out.flush();
         }
-      }
-
-      if (rowWriterThread.isAlive()) {
-        try {
-          rowWriterThread.interrupt();
-          rowWriterThread.join();
-        } catch (InterruptedException exception) {
-          log.warn(
-              "Failed to join row writer thread; setting to daemon to avoid hanging on shutdown"
-          );
-          rowWriterThread.setDaemon(true);
+        if (streamsException != null) {
+          throw streamsException;
         }
       }
+    } catch (EOFException exception) {
+      // The user has terminated the connection; we can stop writing
```

Comment:
Should we at least log this so we know why it has been terminated?

Reply:
done